File: 120-triangle/triangle.cpp
class Solution {
private:
    int help(vector<vector<int>>& dp, int x, int y,
             vector<vector<int>>& triangle) {
        if (x >= triangle.size() - 1)
            return triangle[x][y];
        if (dp[x][y] != INT_MAX)
            return dp[x][y];

        int first = help(dp, x + 1, y, triangle);
        int second = help(dp, x + 1, y + 1, triangle);

        return dp[x][y] = min(first+ triangle[x][y], second+ triangle[x][y]);
    }

public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int m = triangle.size();
        vector<vector<int>> dp(m, vector<int>(m, INT_MAX));
        return help(dp, 0, 0, triangle);
    }
};

File: 120-triangle/triangle.java
class Solution {
    private int[][] memo;
    public int helper(List<List<Integer>> triangle, int x, int y) {
        if (x >= triangle.size()) return 0;
        if (memo[x][y] != Integer.MAX_VALUE) return memo[x][y];

        int temp = triangle.get(x).get(y) +
                Math.min(helper(triangle, x + 1, y), helper(triangle, x + 1, y + 1));

        memo[x][y] = temp;
        return temp;
    }
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        memo = new int[n][n];
        for (int[] row : memo) Arrays.fill(row, Integer.MAX_VALUE);

        return helper(triangle, 0, 0);
    }
}
